sway-core/src/type_system/info.rs

```diff
@@ -1526,6 +1526,246 @@ impl TypeInfo {
+                    .fold(AbiEncodeSizeHint::Exact(0), |old_size_hint, t| {
+                        let field_type = engines.te().get(t.type_id);
+                        let field_size_hint = field_type.abi_encode_size_hint(engines);
+                        old_size_hint + field_size_hint
+                    })
+            }
+
+            TypeInfo::Struct(s) => {
+                let decl = engines.de().get(s.id());
+                decl.fields
+                    .iter()
+                    .fold(AbiEncodeSizeHint::Exact(0), |old_size_hint, f| {
+                        let field_type = engines.te().get(f.type_argument.type_id);
+                        let field_size_hint = field_type.abi_encode_size_hint(engines);
+                        old_size_hint + field_size_hint
+                    })
+            }
+            TypeInfo::Enum(e) => {
+                let decl = engines.de().get(e.id());
+
+                let min = decl
+                    .variants
+                    .iter()
+                    .fold(None, |old_size_hint: Option<AbiEncodeSizeHint>, v| {
+                        let variant_type = engines.te().get(v.type_argument.type_id);
+                        let current_size_hint = variant_type.abi_encode_size_hint(engines);
+                        match old_size_hint {
+                            Some(old_size_hint) => Some(old_size_hint.min(current_size_hint)),
+                            None => Some(current_size_hint),
+                        }
+                    })
+                    .unwrap_or(AbiEncodeSizeHint::Exact(0));
+
+                let max =
+                    decl.variants
+                        .iter()
+                        .fold(AbiEncodeSizeHint::Exact(0), |old_size_hint, v| {
+                            let variant_type = engines.te().get(v.type_argument.type_id);
+                            let current_size_hint = variant_type.abi_encode_size_hint(engines);
+                            old_size_hint.max(current_size_hint)
+                        });
+
+                AbiEncodeSizeHint::range_from_min_max(min, max) + 8
+            }
+
+            x => unimplemented!("abi_encode_size_hint for [{}]", engines.help_out(x)),
+        }
+    }
+}
+
+#[derive(Debug)]
+pub enum AbiEncodeSizeHint {
+    CustomImpl,
+    PotentiallyInfinite,
+    Exact(usize),
+    Range(usize, usize),
+}
+
+impl AbiEncodeSizeHint {
+    fn range_from_min_max(a: AbiEncodeSizeHint, b: AbiEncodeSizeHint) -> AbiEncodeSizeHint {
+        match (a, b) {
+            (AbiEncodeSizeHint::CustomImpl, _) => AbiEncodeSizeHint::CustomImpl,
+            (_, AbiEncodeSizeHint::CustomImpl) => AbiEncodeSizeHint::CustomImpl,
+            (AbiEncodeSizeHint::PotentiallyInfinite, _) => AbiEncodeSizeHint::PotentiallyInfinite,
+            (_, AbiEncodeSizeHint::PotentiallyInfinite) => AbiEncodeSizeHint::PotentiallyInfinite,
+            (AbiEncodeSizeHint::Exact(l), AbiEncodeSizeHint::Exact(r)) => {
```

Comment:
If `a > b` we will get a `Range` whose min is larger then max. I assume we count on valid usage of the function and don't want to return `Result` or swap `a` and `b` or.... Suggestion is to have at least an `assert` here to catch an invalid usage? Same for all the arms that return `Range`.

Reply:
Done. Now all range creation have an assert.